<div style="width=50%"><img src="dtsLogo.png" width="214.5px" height="102px" style="float:left"></img></div>
<div style="width=50%"><img src="unudLogo.png" width="102px" height="102px" style="float:right"></img></div>

# <center>Nutrition Fact for McDonald’s Menu</center>

<center><b><font size=5>Group 4</font></b></center><br>
<center>Andrew Dwi Permana</center>
<center>Eggi Dwitama</center>
<center>Putu Gede Gaura Priajana</center>


## Background
The question based on dataset provided in https://www.kaggle.com/mcdonalds/nutrition-facts. The Dataset is from 2017 and contains nutritional value of 260 MCDonald's menu that available in United States. 

## Goal:
<ol type="a">
    <li>How many calories does the average McDonald's value meal contain?</li> 
    <li>How much do beverages, like soda or coffee, contribute to the overall caloric intake?</li> 
    <li>Does ordered grilled chicken instead of crispy increase a sandwich's nutritional value?</li> 
    <li>What about ordering egg whites instead of whole eggs?</li> 
    <li>What is the least number of items could you order from the menu to meet one day's nutritional requirements?</li>
 </ol>

In [1]:
import matplotlib.pyplot as plt
import numpy as np 
import pandas as pd

The first Step is to look at the dataset closely to have it's rough picture 

In [2]:
df = pd.read_csv('mcdonal_menu_dataset.csv')
df.head()

Category                              Item    Serving Size  Calories  \
0  Breakfast                      Egg McMuffin  4.8 oz (136 g)       300   
1  Breakfast                 Egg White Delight  4.8 oz (135 g)       250   
2  Breakfast                  Sausage McMuffin  3.9 oz (111 g)       370   
3  Breakfast         Sausage McMuffin with Egg  5.7 oz (161 g)       450   
4  Breakfast  Sausage McMuffin with Egg Whites  5.7 oz (161 g)       400   

   Calories from Fat  Total Fat  Total Fat (% Daily Value)  Saturated Fat  \
0                120       13.0                         20            5.0   
1                 70        8.0                         12            3.0   
2                200       23.0                         35            8.0   
3                250       28.0                         43           10.0   
4                210       23.0                         35            8.0   

   Saturated Fat (% Daily Value)  Trans Fat  ...  Carbohydrates  \
0                             25        0.0  ...             31   
1                             15        0.0  ...             30   
2                             42        0.0  ...             29   
3                             52        0.0  ...             30   
4                             42        0.0  ...             30   

   Carbohydrates (% Daily Value)  Dietary Fiber  \
0                             10              4   
1                             10              4   
2                             10              4   
3                             10              4   
4                             10              4   

   Dietary Fiber (% Daily Value)  Sugars  Protein  Vitamin A (% Daily Value)  \
0                             17       3       17                         10   
1                             17       3       18                          6   
2                             17       2       14                          8   
3                             17       2       21                         15   
4                             17       2       21                          6   

   Vitamin C (% Daily Value)  Calcium (% Daily Value)  Iron (% Daily Value)  
0                          0                       25                    15  
1                          0                       25                     8  
2                          0                       25                    10  
3                          0                       30                    15  
4                          0                       25                    10  

[5 rows x 24 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 260 entries, 0 to 259
Data columns (total 24 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Category                       260 non-null    object 
 1   Item                           260 non-null    object 
 2   Serving Size                   260 non-null    object 
 3   Calories                       260 non-null    int64  
 4   Calories from Fat              260 non-null    int64  
 5   Total Fat                      260 non-null    float64
 6   Total Fat (% Daily Value)      260 non-null    int64  
 7   Saturated Fat                  260 non-null    float64
 8   Saturated Fat (% Daily Value)  260 non-null    int64  
 9   Trans Fat                      260 non-null    float64
 10  Cholesterol                    260 non-null    int64  
 11  Cholesterol (% Daily Value)    260 non-null    int64  
 12  Sodium                         260 non-null    int

We see that Category is the very important in answering questions in goal. So let's take a look what kind of Category is there

In [4]:
df['Category'].value_counts()

Coffee & Tea          95
Breakfast             42
Smoothies & Shakes    28
Chicken & Fish        27
Beverages             27
Beef & Pork           15
Snacks & Sides        13
Desserts               7
Salads                 6
Name: Category, dtype: int64

We are now prepared to answer the question, let's go

# a) How many calories does the average McDonald's value meal contain?
The key in this question is <b><i>meal</i></b>. It asked only for meal type menu. We then filter menu based on their category and find its average calories.  

In [5]:
df_meal = df[(df['Category'] == 'Chicken & Fish') | (df['Category'] == 'Beef & Pork' ) | (df['Category'] == 'Breakfast' )]
print('Calories average contain on McDonald\'s value meal:', df_meal["Calories"].mean())

Calories average contain on McDonald's value meal: 529.2857142857143


# b) How much do beverages, like soda or coffee, contribute to the overall caloric intake?
The key in this question is <b><i>Overal caloric intake</i></b>. We saw that there's no column for overall caloric intake, only amount of calories for each menu. So we had to do some research and found that a person needs 2000 calories a day (the value might be different based on age, gender,and daily activities, to simplify calculation we just use 2000). We created a new column that contains percentage of calories of each menu based on 2000 calories requirements then we filter the the dataframe to show only beverages types of menu. 

In [6]:
#create new column for % calories intake 
#daily calories intake is 2000 calories
df['Calories (% Daily Value)'] = (df['Calories']/2000)* 100

In [7]:
#filter dataframe
df_beverage = df.query('Category == "Coffee & Tea" | Category == "Smoothies & Shakes" | Category == "Beverages"') 
df_beverage = df_beverage.loc[:,['Category','Item','Calories','Calories (% Daily Value)']]
df_beverage.sort_values('Calories (% Daily Value)',ascending=False,inplace=True)
df_beverage

# #plot result
# x=df_beverage['Item']
# y=df_beverage['Calories (% Daily Value)']
# fig, ax = plt.subplots()
# ax.barh(x, y)

Category                                  Item  Calories  \
253  Smoothies & Shakes  McFlurry with M&M’s Candies (Medium)       930   
246  Smoothies & Shakes              Strawberry Shake (Large)       850   
249  Smoothies & Shakes               Chocolate Shake (Large)       850   
243  Smoothies & Shakes                 Vanilla Shake (Large)       820   
251  Smoothies & Shakes                Shamrock Shake (Large)       820   
..                  ...                                   ...       ...   
140        Coffee & Tea                      Iced Tea (Child)         0   
122           Beverages                Diet Dr Pepper (Small)         0   
123           Beverages               Diet Dr Pepper (Medium)         0   
124           Beverages                Diet Dr Pepper (Large)         0   
125           Beverages                Diet Dr Pepper (Child)         0   

     Calories (% Daily Value)  
253                      46.5  
246                      42.5  
249                      42.5  
243                      41.0  
251                      41.0  
..                        ...  
140                       0.0  
122                       0.0  
123                       0.0  
124                       0.0  
125                       0.0  

[150 rows x 4 columns]

In the table above shown 5 top most beverages and 5 lower most beverage sorted by their contribution to daily caloric requirement

# c) Does ordered grilled chicken instead of crispy increase a sandwich's nutritional value?
The key in this question is <i><b>nutritional value</b></i> but there's no column explicitly shown it. After researching we come to conclusion that each column represent a part of nutritional value. We can represent nutritional value by summing up each column (beside columns that represent percentage (%) beause it use different scale). Then We filter for only menu that has "sandwich" in it and present it side by side to compare it easily. In case there's a menu that have no pair, we drop it. 

In [8]:
#check which value is higher
def HigherNutValue (row):
   if row['Total Nutritional Grilled'] > row['Total Nutritional Crispy'] :
      return 'GRILLED'
   else :
      return 'CRISPY'

#select relevant data
df_sandwich = df[df['Item'].str.contains("Sandwich")]

#find total nutritional value
df_buffer= df_sandwich.loc[:,['Item','Calories','Total Fat','Trans Fat','Sodium','Carbohydrates','Dietary Fiber','Sugars','Protein']]
df_buffer['Total Nutritional'] = df_buffer.sum(axis=1)
df_buffer

#remove unwanted data
df_buffer.drop(labels=65, axis=0, inplace=True) 
df_buffer.drop(labels=['Calories','Total Fat','Trans Fat','Sodium','Carbohydrates','Dietary Fiber','Sugars','Protein'],axis=1,inplace=True)

#split 
df_grilled = df_buffer[df_buffer['Item'].str.contains("Grilled Chicken")]
df_crispy = df_buffer[df_buffer['Item'].str.contains("Crispy Chicken")]

#rename splitted & reset index
df_grilled.columns=('Item Grilled', 'Total Nutritional Grilled')
df_grilled.reset_index(drop=True, inplace=True)

df_crispy.columns=('Item Crispy', 'Total Nutritional Crispy')
df_crispy.reset_index(drop=True, inplace=True)

#combine table
df_compare = pd.concat([df_grilled, df_crispy],axis=1, join='inner')
df_compare['HigherNutValue'] = df_compare.apply (lambda row: HigherNutValue(row), axis=1)
df_compare.head()

Item Grilled  Total Nutritional Grilled  \
0    Premium Grilled Chicken Classic Sandwich                     1260.0   
1       Premium Grilled Chicken Club Sandwich                     1876.0   
2  Premium Grilled Chicken Ranch BLT Sandwich                     1786.0   
3    Bacon Clubhouse Grilled Chicken Sandwich                     2284.0   

                                 Item Crispy  Total Nutritional Crispy  \
0    Premium Crispy Chicken Classic Sandwich                    1614.0   
1       Premium Crispy Chicken Club Sandwich                    2221.0   
2  Premium Crispy Chicken Ranch BLT Sandwich                    2141.0   
3    Bacon Clubhouse Crispy Chicken Sandwich                    2629.5   

  HigherNutValue  
0         CRISPY  
1         CRISPY  
2         CRISPY  
3         CRISPY

The table above shows sandwich menu that contain Crispy Chicken and Grilled Chicked side by side along with its nutritional value. We also made some little function that automatically compare both and return which one has higher nutritional value. From the four pair menu compared, all sandwich that use Crispy chicken give more nutritional value

# d) What about ordering egg whites instead of whole eggs?
This question is pretty similar with the previous question, hence our approach is the same.

In [9]:
#check which value is higher
def HigherNutValue (row):
   if row['Total Nutritional Egg White'] > row['Total Nutritional Egg Whole'] :
      return 'Egg White'
   else :
      return 'Egg Whole'

df_egg = df[df['Item'].str.contains("Egg")]

# find total nutritional value
df_buffer2= df_egg.loc[:,['Item','Calories','Total Fat','Trans Fat','Sodium','Carbohydrates','Dietary Fiber','Sugars','Protein']]
df_buffer2['Total Nutritional'] = df_buffer2.sum(axis=1)
df_buffer2.reset_index(drop=True,inplace=True)

#remove unwanted data
df_buffer2.drop(df_buffer2.index[[4,13,20,21,22,23,24]], axis=0, inplace=True) 
df_buffer2.drop(labels=['Calories','Total Fat','Trans Fat','Sodium','Carbohydrates','Dietary Fiber','Sugars','Protein'],axis=1,inplace=True)

#split 
df_eggWhite = df_buffer2[df_buffer2['Item'].str.contains("Egg White")]
df_eggWhole = df_buffer2[~df_buffer2['Item'].str.contains("Egg White")]

#rename splitted & reset index
df_eggWhite.columns=('Item Egg White', 'Total Nutritional Egg White')
df_eggWhite.reset_index(drop=True, inplace=True)

df_eggWhole.columns=('Item Egg Whole', 'Total Nutritional Egg Whole')
df_eggWhole.reset_index(drop=True, inplace=True)

#combine table
df_compare = pd.concat([df_eggWhite, df_eggWhole],axis=1)
df_compare['HigherNutValue'] = df_compare.apply (lambda row: HigherNutValue(row), axis=1)
df_compare.head(15)

Item Egg White  \
0                                  Egg White Delight   
1                   Sausage McMuffin with Egg Whites   
2  Bacon, Egg & Cheese Biscuit with Egg Whites (R...   
3  Bacon, Egg & Cheese Biscuit with Egg Whites (L...   
4  Sausage Biscuit with Egg Whites (Regular Biscuit)   
5    Sausage Biscuit with Egg Whites (Large Biscuit)   
6     Bacon, Egg & Cheese McGriddles with Egg Whites   
7   Sausage, Egg & Cheese McGriddles with Egg Whites   
8          Bacon, Egg & Cheese Bagel with Egg Whites   

   Total Nutritional Egg White                                 Item Egg Whole  \
0                       1083.0                                   Egg McMuffin   
1                       1360.0                      Sausage McMuffin with Egg   
2                       1791.0  Bacon, Egg & Cheese Biscuit (Regular Biscuit)   
3                       1984.0    Bacon, Egg & Cheese Biscuit (Large Biscuit)   
4                       1724.0     Sausage Biscuit with Egg (Regular Biscuit)   
5                       1906.0       Sausage Biscuit with Egg (Large Biscuit)   
6                       1750.0                 Bacon, Egg & Cheese McGriddles   
7                       1930.0               Sausage, Egg & Cheese McGriddles   
8                       2171.5                      Bacon, Egg & Cheese Bagel   

   Total Nutritional Egg Whole HigherNutValue  
0                       1118.0      Egg Whole  
1                       1395.0      Egg Whole  
2                       1848.0      Egg Whole  
3                       2029.0      Egg Whole  
4                       1771.0      Egg Whole  
5                       1953.0      Egg Whole  
6                       1815.0      Egg Whole  
7                       1986.0      Egg Whole  
8                       2228.5      Egg Whole

The table above shows menu that contain White Egg and Whole Egg side by side along with its nutritional value. From the eight pair menu compared, all menu that use Whole Egg give more nutritional value

# e) What is the least number of items could you order from the menu to meet one day's nutritional requirements?
the key in this question is <i><b>least number of items</b></i> to fulfill daily nutritional requirements. For this question we use only column that is in percetage. Reason being is the percetage is already based on daily value requirements. Next we take the the top menu from each percentage column that already sorted. We then sum up each column to see whether all column have value >= 100% which means nutritional requirements for that column already fulfilled. 

In [10]:
#remove unwanted column
df_buffer3= df.loc[:,['Item','Total Fat (% Daily Value)','Sodium (% Daily Value)','Carbohydrates (% Daily Value)','Dietary Fiber (% Daily Value)','Vitamin A (% Daily Value)', 'Vitamin C (% Daily Value)','Calcium (% Daily Value)', 'Iron (% Daily Value)','Calories (% Daily Value)']]

#create empty dataframe
df_buffer4 = df_buffer3
df_buffer4 = pd.DataFrame(columns=df_buffer4.columns)

# add the top menu from each column
df_buffer3.sort_values('Total Fat (% Daily Value)', ascending=False,inplace=True)
df_buffer4 = df_buffer4.append(df_buffer3.head(1))

df_buffer3.sort_values('Sodium (% Daily Value)', ascending=False,inplace=True)
df_buffer4 = df_buffer4.append(df_buffer3.head(1))

df_buffer3.sort_values('Carbohydrates (% Daily Value)', ascending=False,inplace=True)
df_buffer4 = df_buffer4.append(df_buffer3.head(1))

df_buffer3.sort_values('Dietary Fiber (% Daily Value)', ascending=False,inplace=True)
df_buffer4 = df_buffer4.append(df_buffer3.head(1))

df_buffer3.sort_values('Vitamin A (% Daily Value)', ascending=False,inplace=True)
df_buffer4 = df_buffer4.append(df_buffer3.head(1))

df_buffer3.sort_values('Vitamin C (% Daily Value)', ascending=False,inplace=True)
df_buffer4 = df_buffer4.append(df_buffer3.head(1))

df_buffer3.sort_values('Calcium (% Daily Value)', ascending=False,inplace=True)
df_buffer4 = df_buffer4.append(df_buffer3.head(1))

df_buffer3.sort_values('Iron (% Daily Value)', ascending=False,inplace=True)
df_buffer4 = df_buffer4.append(df_buffer3.head(1))

df_buffer3.sort_values('Calories (% Daily Value)', ascending=False,inplace=True)
df_buffer4 = df_buffer4.append(df_buffer3.head(1))

#drop duplicate data
df_buffer4.drop_duplicates(subset=['Item'],inplace=True)
df_buffer4

# check manually if sum of each column is >100
df_buffer5 = df_buffer4.sum(axis=0)
print(df_buffer5)

Item                             Chicken McNuggets (40 piece)Chocolate Shake (L...
Total Fat (% Daily Value)                                                      394
Sodium (% Daily Value)                                                         306
Carbohydrates (% Daily Value)                                                  207
Dietary Fiber (% Daily Value)                                                   95
Vitamin A (% Daily Value)                                                      240
Vitamin C (% Daily Value)                                                      287
Calcium (% Daily Value)                                                        187
Iron (% Daily Value)                                                           105
Calories (% Daily Value)                                                       277
dtype: object


Above shown the sum of each percentage column for the 6 menu that give the most value for one or more column. We still lack dietary fiber by 5%. We need at least 1 more menu to fulfill it. So our conclusion is: it takes at least 7 menu to fulfill daily nutritional requirements

## Conclusion:
<ol type="a">
    <li>How many calories does the average McDonald's value meal contain? <b>529 Calories</b> </li> 
    <li>How much do beverages, like soda or coffee, contribute to the overall caloric intake? <b>For a drink at most 46.5% of daily caloric requirements</b> </li> 
    <li>Does ordered grilled chicken instead of crispy increase a sandwich's nutritional value? <b>Crispy chicken give more nutritional value for sandwich</b></li> 
    <li>What about ordering egg whites instead of whole eggs? <b>Whole Egg give more nutritional value</b></li> 
    <li>What is the least number of items could you order from the menu to meet one day's nutritional requirements? <b>A person need to consume at least 7 different MCDonalds menu to fullfill his daily nutrional requirements</b></li>
 </ol>